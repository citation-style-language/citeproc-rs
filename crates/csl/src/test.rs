// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Copyright Â© 2020 Corporation for Digital Scholarship

use super::*;

#[test]
fn features() {
    assert_snapshot_parse!(Features, r#"<features></features>"#);
    assert_snapshot_parse!(
        Features,
        r#"
        <features>
            <feature name="condition-date-parts" />
            <feature name="edtf-dates" />
        </features>
    "#
    );
    assert_snapshot_err!(
        Features,
        r#"
        <features>
            <feature name="edtf-dates" />
            <feature name="UNRECOGNIZED-FEATURE" />
            <feature name="SECOND-UNRECOGNIZED-FEATURE" />
        </features>
    "#
    );
}

#[test]
fn intext() {
    let features = Features {
        custom_intext: true,
        ..Default::default()
    };
    let options = ParseOptions {
        allow_no_info: true,
        features: Some(features),
        ..Default::default()
    };
    assert_snapshot_parse!(
        InText,
        r#"<intext><layout><text variable="title"/></layout></intext>"#
    );
    assert_snapshot_parse!(
        Style,
        r#"<style class="in-text">
            <citation><layout></layout></citation>
            <intext><layout><text variable="title" /></layout></intext>
        </style>"#,
        options.clone()
    );
    assert_snapshot_err!(
        Style,
        r#"<style class="in-text">
             <citation><layout></layout></citation>
             <intext><layout></layout></intext>
         </style>"#
    );
}

#[test]
fn unsupported_version() {
    assert_snapshot_err!(
        Style,
        r#"
        <style version="999.0" class="in-text">
            <citation><layout></layout></citation>
        </style>
    "#
    );
}

#[test]
fn unrecognised_macros() {
    assert_snapshot_err!(
        Style,
        r#"
        <style version="1.0" class="in-text">
            <citation>
                <layout>
                    <text macro="unknown" />
                </layout>
            </citation>
        </style>
    "#
    );
    assert_snapshot_err!(
        Style,
        r#"
        <style version="1.0" class="in-text">
            <citation>
                <sort>
                    <key macro="unknown" />
                </sort>
                <layout></layout>
            </citation>
        </style>
    "#
    );
    assert_snapshot_err!(
        Style,
        r#"
        <style version="1.0" class="in-text">
            <citation><layout></layout></citation>
            <bibliography>
                <sort>
                    <key macro="unknown" />
                </sort>
                <layout></layout>
            </bibliography>
        </style>
    "#
    );
    assert_snapshot_parse!(
        Style,
        r#"
        <style version="1.0" class="in-text">
            <macro name="known" />
            <citation>
                <layout>
                    <text macro="known" />
                </layout>
            </citation>
        </style>
    "#
    );
}

#[test]
fn missing_info() {
    // Externally, missing info should fail.
    insta::assert_debug_snapshot!(crate::Style::parse(::indoc::indoc!(
        r#"
            <style version="1.0.1" class="in-text">
                <citation><layout></layout></citation>
            </style>
        "#
    ))
    .expect_err("should have failed with errors"));
    // But internally we can ignore it.
    assert_snapshot_parse!(
        Style,
        r#"
        <style version="1.0.1" class="in-text">
            <citation><layout></layout></citation>
        </style>
    "#
    );
}

#[test]
fn wrong_tag_name() {
    assert_snapshot_err!(
        Style,
        r#"
        <stylo version="1.0.1" class="in-text">
            <citation><layout></layout></citation>
        </stylo>
    "#
    );
    assert_snapshot_err!(
        Locale,
        r#"
        <localzzz xml:lang="en-US" version="1.0.1" class="in-text">
        </localzzz>
    "#
    );
}
