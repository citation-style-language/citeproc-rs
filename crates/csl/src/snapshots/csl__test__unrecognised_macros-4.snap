---
source: crates/csl/src/test.rs
expression: "crate::Style::parse_for_test(::indoc::indoc!(r#\"\n        <style version=\"1.0\" class=\"in-text\">\n            <macro name=\"known\" />\n            <citation>\n                <layout>\n                    <text macro=\"known\" />\n                </layout>\n            </citation>\n        </style>\n    \"#)).expect(\"should have parsed successfully\")"
---
Style {
    class: InText,
    macros: {
        "known": [],
    },
    citation: Citation {
        disambiguate_add_names: false,
        disambiguate_add_givenname: false,
        givenname_disambiguation_rule: ByCite,
        disambiguate_add_year_suffix: false,
        layout: Layout {
            affixes: None,
            formatting: None,
            delimiter: None,
            elements: [
                Text(
                    TextElement {
                        source: Macro(
                            "known",
                        ),
                        formatting: None,
                        affixes: None,
                        quotes: false,
                        strip_periods: false,
                        text_case: None,
                        display: None,
                    },
                ),
            ],
            locale: [],
        },
        name_inheritance: Name {
            and: None,
            delimiter: None,
            delimiter_precedes_et_al: None,
            delimiter_precedes_last: None,
            et_al_min: None,
            et_al_use_first: None,
            et_al_use_last: None,
            et_al_subsequent_min: None,
            et_al_subsequent_use_first: None,
            form: None,
            initialize: None,
            initialize_with: None,
            name_as_sort_order: None,
            sort_separator: None,
            formatting: None,
            affixes: None,
            name_part_given: None,
            name_part_family: None,
        },
        names_delimiter: None,
        near_note_distance: 5,
        sort: None,
        cite_group_delimiter: None,
        year_suffix_delimiter: None,
        after_collapse_delimiter: None,
        collapse: None,
    },
    bibliography: None,
    intext: None,
    info: Info {
        id: Identifier(
            "default",
        ),
        updated: 2000-01-01T00:00:00+00:00,
        title: LocalizedString {
            value: "",
            lang: None,
        },
        title_short: None,
        rights: None,
        summary: None,
        parent: None,
        links: [],
        citation_format: None,
        categories: [],
        issn: None,
        eissn: None,
        issnl: None,
    },
    features: Features {},
    name_inheritance: Name {
        and: None,
        delimiter: None,
        delimiter_precedes_et_al: None,
        delimiter_precedes_last: None,
        et_al_min: None,
        et_al_use_first: None,
        et_al_use_last: None,
        et_al_subsequent_min: None,
        et_al_subsequent_use_first: None,
        form: None,
        initialize: None,
        initialize_with: None,
        name_as_sort_order: None,
        sort_separator: None,
        formatting: None,
        affixes: None,
        name_part_given: None,
        name_part_family: None,
    },
    names_delimiter: None,
    locale_overrides: {},
    default_locale: None,
    version_req: CslVersionReq(
        VersionReq {
            ranges: [
                Range {
                    predicates: [
                        Predicate {
                            op: GtEq,
                            major: 1,
                            minor: 0,
                            patch: 0,
                            pre: [],
                        },
                        Predicate {
                            op: Lt,
                            major: 1,
                            minor: 1,
                            patch: 0,
                            pre: [],
                        },
                    ],
                    compat: Cargo,
                },
            ],
            compat: Cargo,
        },
    ),
    page_range_format: None,
    demote_non_dropping_particle: DisplayAndSort,
    initialize_with_hyphen: true,
}
